retworkx-core/src/centrality.rs

```diff
@@ -297,3 +300,201 @@ where
+/// is smaller than the error tolerance multiplied by the number of nodes.
+/// The implementation of this algorithm is based on the NetworkX
+/// [`eigenvector_centrality()`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.centrality.eigenvector_centrality.html)
+/// function.
+///
+/// In the case of multigraphs the weights of any parallel edges will be
+/// summed when computing the eigenvector centrality.
+///
+/// Arguments:
+///
+/// * `graph` - The graph object to run the algorithm on
+/// * `weight_fn` - An input callable that will be pased the `EdgeRef` for
+///     an edge in the graph and is expected to return a `Result<f64>` of
+///     the weight of that edge.
+/// * `max_iter` - The maximum number of iterations in the power method. If
+///     set to `None` a default value of 100 is used.
+/// * `tol` - The error tolerance used when checking for convergence in the
+///     power method. If set to `None` a dfault value of 1e-6 is used.
+///
+/// # Example
+/// ```rust
+/// use hashbrown::HashMap;
+///
+/// use retworkx_core::Result;
+/// use retworkx_core::petgraph;
+/// use retworkx_core::petgraph::visit::{IntoEdges, IntoNodeIdentifiers};
+/// use retworkx_core::centrality::eigenvector_centrality;
+///
+/// let g = petgraph::graph::UnGraph::<i32, ()>::from_edges(&[
+///     (0, 1), (1, 2)
+/// ]);
+/// // Calculate the betweeness centrality
+/// let output: Result<Option<HashMap<petgraph::graph::NodeIndex, f64>>> =
+///     eigenvector_centrality(&g, |_| {Ok(1.)}, None, None);
+/// ```
+pub fn eigenvector_centrality<G, F, E>(
+    graph: G,
+    mut weight_fn: F,
+    max_iter: Option<usize>,
+    tol: Option<f64>,
+) -> Result<Option<HashMap<G::NodeId, f64>>, E>
+where
+    G: NodeIndexable
+        + IntoNodeIdentifiers
+        + IntoNeighbors
+        + IntoEdges
+        + NodeCount
+        + GraphProp
+        + GraphBase<NodeId = NodeIndex>,
+    F: FnMut(G::EdgeRef) -> Result<f64, E>,
+{
+    let tol: f64 = tol.unwrap_or(1e-6);
+    let max_iter = max_iter.unwrap_or(50);
+    let n_start: HashMap<G::NodeId, f64> = graph.node_identifiers().map(|n| (n, 1.)).collect();
+    let n_start_sum: f64 = n_start.len() as f64;
+    let mut x: HashMap<G::NodeId, f64> =
+        n_start.iter().map(|(k, v)| (*k, v / n_start_sum)).collect();
+    let node_count = graph.node_count();
+    for _ in 0..max_iter {
+        let x_last = x.clone();
+        for node in x_last.keys() {
+            for neighbor in graph.neighbors(*node) {
+                let w_vec: Vec<G::EdgeRef> = graph
+                    .edges(*node)
+                    .filter(|edge| edge.target() == neighbor)
+                    .collect();
+                let mut w = 0.;
+                for edge in w_vec {
+                    w += weight_fn(edge)?;
+                }
+                *x.get_mut(&neighbor).unwrap() += x_last[node] * w;
+            }
```

Comment:
I'm not sure if this is correct in case of parallel edges: `graph.neighbors` will return a neighbor as many times as the number of parallel edges and each time we sum the weights of all parallel edges, so we are "overcounting".

```suggestion
            for edge in graph.edges(*node) {
                let w = weight_fn(edge)?;
                let neighbor = edge.target();
                *x.get_mut(&neighbor).unwrap() += x_last[node] * w;
            }
```

Reply:
Well my thinking was for parallel edges the centrality would be higher so it worked out ok. Similarly if we were going to compute this by generating an adjacency matrix and finding the eigenvector for it we'd be doing a similar thing because the adjacency matrix sums the weight of parallel edges.
'
That being said I don't have a super strong opinion on the correct behavior here, so I'll defer to whatever you think is best

Reply:
Yeah, I agree. It's the right approach to sum the weight of parallel edges but my point was that the total update of the above loop is: `x[neighbor] += x_last[node] * sum(weights of parallel (node, neighbor) edges) * nb. of parallel (node, neighbor) edges` which is incorrect.

Reply:
Ah, my bad I misread your comment. Yeah that makes sense. I'll correct this

Reply:
Fixed in: https://github.com/Qiskit/retworkx/pull/621/commits/e03856b1f6ee4430d49f45e915e1683074d54884